fuel-relayer/src/finalization_queue.rs

```diff
@@ -0,0 +1,521 @@
+                    .zip(amounts.iter())
+                    .map(|(f, s)| (*f, *s))
+                    .collect();
+                last_diff.delegations.insert(delegator, Some(delegates));
+            }
+            EthEventLog::ValidatorRegistration {
+                staking_key,
+                consensus_key,
+            } => {
+                last_diff
+                    .validators
+                    .insert(staking_key, Some(consensus_key));
+            }
+            EthEventLog::ValidatorUnregistration { staking_key } => {
+                last_diff.validators.insert(staking_key, None);
+            }
+            EthEventLog::FuelBlockCommited { height, block_root } => {
+                self.blocks
+                    .handle_block_commit(block_root, (height).into(), da_height, false);
+            }
+            EthEventLog::Unknown => (),
+        }
+    }
+
+    /// Used to commit da block diff to database.
+    pub async fn commit_diffs(
+        &mut self,
+        db: &mut dyn RelayerDb,
+        finalized_da_height: DaBlockHeight,
+    ) {
+        if self.finalized_da_height >= finalized_da_height {
+            error!(
+                "We received finalized height {} but we already have {}",
+                finalized_da_height, self.finalized_da_height
+            );
+            return;
+        }
+        while let Some(diff) = self.pending.front_mut() {
+            if diff.da_height > finalized_da_height {
+                break;
+            }
+            info!("flush eth log:{:?} diff:{:?}", diff.da_height, diff);
+            //TODO to be paranoid, recheck events got from eth client.
+
+            // apply staking diffs
+            db.insert_staking_diff(
+                diff.da_height,
+                &StakingDiff::new(diff.validators.clone(), diff.delegations.clone()),
+            )
+            .await;
+
+            // append index of delegator so that we cross reference earliest delegation set
+            for (delegate, _) in diff.delegations.iter() {
+                db.append_delegate_index(delegate, diff.da_height).await;
+            }
+
+            // push finalized assets to db
+            for (_, deposit) in diff.assets.iter() {
+                db.insert_coin_deposit(deposit.into()).await
+            }
+
+            // insert height index into delegations.
+            db.set_finalized_da_height(diff.da_height).await;
+
+            // remove pending diff
+            self.pending.pop_front();
+        }
+
+        let last_commited_fin_fuel_height = self.blocks.handle_da_finalization(finalized_da_height);
+
+        db.set_last_commited_finalized_fuel_height(last_commited_fin_fuel_height)
+            .await;
+        self.finalized_da_height = finalized_da_height;
+        // bump validator set to last finalized block
+        self.validators
+            .bump_set_to_da_height(finalized_da_height, db)
+            .await
+    }
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
There are no tests involving removed logs (reorgs) while there is a lot of complex logic above to deal with them.

Reply:
Will add tests